### ISSUE TITLE
ABI decode fails mapping bytes to DynamicStruct

### ISSUE THREAD

**Issue by denyshorman** [web3j/web3j#1426]
Adding tests that reproduce the problem

```java
public static class Entity extends DynamicStruct {
    public String param1;
    public String param2;
    public BigInteger param3;
    public BigInteger param4;
    public BigInteger param5;
    public List<String> param6;
    public String param7;
    public List<BigInteger> param8;
    public List<BigInteger> param9;
    public String param10;
    public String param11;
    public Boolean param12;

    Entity(
            String param1,
            String param2,
            BigInteger param3,
            BigInteger param4,
            BigInteger param5,
            List<String> param6,
            String param7,
            List<BigInteger> param8,
            List<BigInteger> param9,
            String param10,
            String param11,
            Boolean param12
    ) {
        super(
                new Address(param1),
                new Address(param2),
                new Uint256(param3),
                new Uint256(param4),
                new Uint256(param5),
                new DynamicArray<>(Address.class, param6.stream().map(Address::new).collect(Collectors.toList())),
                new DynamicBytes(Numeric.hexStringToByteArray(param7)),
                new DynamicArray<>(Uint256.class, param8.stream().map(Uint256::new).collect(Collectors.toList())),
                new DynamicArray<>(Uint256.class, param9.stream().map(Uint256::new).collect(Collectors.toList())),
                new Address(param10),
                new Utf8String(param11),
                new Bool(param12)
        );

        this.param1 = param1;
        this.param2 = param2;
        this.param3 = param3;
        this.param4 = param4;
        this.param5 = param5;
        this.param6 = param6;
        this.param7 = param7;
        this.param8 = param8;
        this.param9 = param9;
        this.param10 = param10;
        this.param11 = param11;
        this.param12 = param12;
    }

    Entity(
            Address param1,
            Address param2,
            Uint256 param3,
            Uint256 param4,
            Uint256 param5,
            DynamicArray<Address> param6,
            DynamicBytes param7,
            DynamicArray<Uint256> param8,
            DynamicArray<Uint256> param9,
            Address param10,
            Utf8String param11,
            Bool param12
    ) {
        super(
                param1,
                param2,
                param3,
                param4,
                param5,
                param6,
                param7,
                param8,
                param9,
                param10,
                param11,
                param12
        );

        this.param1 = param1.getValue();
        this.param2 = param2.getValue();
        this.param3 = param3.getValue();
        this.param4 = param4.getValue();
        this.param5 = param5.getValue();
        this.param6 = param6.getValue().stream().map(Address::getValue).collect(Collectors.toList());
        this.param7 = Numeric.toHexString(param7.getValue());
        this.param8 = param8.getValue().stream().map(Uint256::getValue).collect(Collectors.toList());
        this.param9 = param9.getValue().stream().map(NumericType::getValue).collect(Collectors.toList());
        this.param10 = param10.getValue();
        this.param11 = param11.getValue();
        this.param12 = param12.getValue();
    }
}

@Test
void testDecodeDynamicStruct_4_8_4() {
    String encoded = "000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec7000000000000000000000000000000000000000000000000000006f62f30dee30000000000000000000000000000000000000000000000000000011f71b7520c000000000000000000000000000000000000000000000000000007dc7add9ced000000000000000000000000000000000000000000000000000000000000018000000000000000000000000000000000000000000000000000000000000001c000000000000000000000000000000000000000000000000000000000000002000000000000000000000000000000000000000000000000000000000000000260000000000000000000000000000000000000000000000000000000000000000100000000000000000000000000000000000000000000000000000000000002a000000000000000000000000000000000000000000000000000000000000000010000000000000000000000000000000000000000000000000000000000000001000000000000000000000000dac17f958d2ee523a2206206994597c13d831ec70000000000000000000000000000000000000000000000000000000000000001110000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000006f0000000000000000000000000000000000000000000000000000000000000001000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000047465737400000000000000000000000000000000000000000000000000000000";

    Function func = new Function(
            "testFunc",
            Collections.emptyList(),
            Collections.singletonList(TypeReference.create(Entity.class))
    );

    // Decode fails with "String index out of range: -2064005070"
    // Reproduced on Web3j 4.8.4 and latest master
    FunctionReturnDecoder.decode(encoded, func.getOutputParameters());
}

@Test
void testDecodeDynamicStructMaster() {
    Entity entity = new Entity(
            "0xdac17f958d2ee523a2206206994597c13d831ec7",
            "0xdac17f958d2ee523a2206206994597c13d831ec7",
            new BigInteger("7654423453411"),
            new BigInteger("1234563453452"),
            new BigInteger("8643535543533"),
            Collections.singletonList("0xdac17f958d2ee523a2206206994597c13d831ec7"),
            "0x11",
            Arrays.asList(new BigInteger("0"), new BigInteger("111")),
            Collections.singletonList(new BigInteger("0")),
            "0x0000000000000000000000000000000000000001",
            "test",
            true
    );

    String encoded = TypeEncoder.encodeDynamicStruct(entity);

    // Decode fails with "Array types must be wrapped in a TypeReference"
    // Reproduced on the latest master branch (75e3c8fba041003eb640156a049806b17b4e52c0)
    TypeDecoder.decodeDynamicStruct(encoded, 0, TypeReference.create(Entity.class));
}
```

**Comment by 0xean** [comment]
Experiencing the same issue.

**Comment by andrii-kl** [comment]
Hi @denyshorman
Thank you for reporting an issue, this problem was fixed and will be released in the next release (4.9.2).
PR  https://github.com/web3j/web3j/pull/1673
